---
title: "1_Chicken_Eye_R_Setup.Rmd"
output: html_document
---

# Chicken Eye: Setup 

Date: July 14 2025

Author: Ben Zazycki

Adapted from: Jared Tangeman

Professor: Dr. Chun Liang

## Installing Packages

Note 1: Only run the cells to install the packages you haven't installed yet.

Note 2: Check the outputs for the install lines. If it's not being installed into your specified libPath from above, ensure that you ran that line. If it's still not working, you can specify `'lib = /path/to/lib'` after the package name.

```{r}
install.packages('remotes')
```

```{r}
install.packages('devtools')
```

```{r}
remotes::install_github('satijalab/seurat-data')
```

```{r}
install.packages('hdf5r')
```

```{r}
remotes::install_github('10XGenomics/loupeR')
```

```{r}
devtools::install_github('immunogenomics/presto')
```

```{r}
install.packages('BiocManager')
```

```{r}
install.packages('reticulate')
```

```{r}
install.packages('Seurat')
```

```{r}
install.packages('Signac')
```

```{r}
install.packages('ggpubr')
```

```{r}
install.packages('ggplot2')
```

```{r}
install.packages('future')
```

```{r}
install.packages('DT')
```

```{r}
install.packages('gprofiler2')
```

```{r}
install.packages('Matrix')
```

```{r}
install.packages('plotly')
```

```{r}
install.packages('ggforce')
```

```{r}
install.packages('patchwork')
```

```{r}
install.packages('presto')
```

```{r}
install.packages('scCustomize')
```

```{r}
install.packages('R.utils')
```

```{r}
BiocManager::install('glmGamPoi')
```

```{r}
BiocManager::install('ensembldb')
```

```{r}
BiocManager::install('JASPAR2024')
```

```{r}
BiocManager::install('TFBSTools')
```

```{r}
BiocManager::install('motifmatchr')
```

```{r}
BiocManager::install('chromVAR')
```

```{r}
BiocManager::install('GenomicRanges')
```

```{r}
BiocManager::install('BSgenomeForge')
```

```{r}
BiocManager::install('biovizBase')
```

```{r}
BiocManager::install('DirichletMultinomial')
```

```{r}
BiocManager::install('GenomeInfoDb')
```

```{r}
BiocManager::install('BSgenome')
```

```{r}
BiocManager::install('Biostrings')
```

```{r}
BiocManager::install('rtracklayer')
```

## Loading from Libraries

Load the packages necessary for this setup document:

```{r}
library(R.utils)
library(Biostrings)
library(rtracklayer)
library(GenomeInfoDb)
library(BSgenome)
library(BSgenomeForge)
```

## Forging Custom Package

In this section, I will demonstrate creating a custom BSgenomeForge package for the relevant genome. This section involves a lot of work with filepaths. Be sure you change everything to your setup on your local machine.

First, make a directory to work in:

```{r}
wd_path <- '/Users/benzazycki/Documents/Research/ChickenEye/GRCg7b.110'
dir.create(wd_path)
```

Download and unzip genome sequence file from ENSEMBL website:

```{r}
ensembl_url <- 'https://ftp.ensembl.org/pub/release-110/fasta/gallus_gallus/dna/Gallus_gallus.bGalGal1.mat.broiler.GRCg7b.dna.toplevel.fa.gz'
destfile_path <- '/Users/benzazycki/Documents/Research/ChickenEye/GRCg7b.110/Gallus_gallus.GRCg7b.fa.gz'
download.file(ensembl_url, destfile = destfile_path)
R.utils::gunzip(destfile_path, remove = FALSE)
```

Read FASTA as DNAStringSet object:

```{r}
fasta_path <- '/Users/benzazycki/Documents/Research/ChickenEye/GRCg7b.110/Gallus_gallus.GRCg7b.fa'
genome_seq <- readDNAStringSet(fasta_path)
```

Export to .2bit format:

```{r}
twobit_file <- '/Users/benzazycki/Documents/Research/ChickenEye/GRCg7b.110/Gallus_gallus.GRCg7b.2bit'
export(genome_seq, twobit_file, format = "2bit")
```

Get sequence info (lengths) from 2bit file:

```{r}
two_bit <- TwoBitFile(twobit_file)
seqinfo_data <- seqinfo(two_bit)
```

Create SeqInfo object with explicit genome name. Save for BSGenome.

```{r}
seqinfo_obj <- Seqinfo(seqnames = seqnames(seqinfo_data),
  seqlengths = seqlengths(seqinfo_data), genome = "GRCg7b")
seq_info_path <- '/Users/benzazycki/Documents/Research/ChickenEye/GRCg7b.110/seqinfo.rds'
saveRDS(seqinfo_obj, seq_info_path)
```

Prepare seed file:

```{r}
seed_path <- '/Users/benzazycki/Documents/Research/ChickenEye/GRCg7b.110/seed.txt'
seqnames_data <- seqnames(seqinfo_data)
seed_content <- sprintf(
"Package: BSgenome.Ggallus.ensembl.GRCg7b
Title: Gallus gallus (GRCg7b) from ENSEMBL
Description: Full genome from ENSEMBL release 110
Version: 1.0.0
organism: Gallus gallus
common_name: Chicken
provider: ENSEMBL
provider_version: GRCg7b
release_date: 2023
source_url: ftp://ftp.ensembl.org/pub/release-110/fasta/gallus_gallus/dna/
seqs_srcdir: /Users/benzazycki/Documents/Research/ChickenEye/GRCg7b.110
seqfile_name: Gallus_gallus.GRCg7b.2bit
seqnames: c(%s)
circ_seqs: 'MT'
genome: GRCg7b
BSgenomeObjname: Ggallus
organism_biocview: Gallus_gallus",
paste(shQuote(seqnames_data, type = "cmd"), collapse = ", "))
writeLines(seed_content, seed_path)
```

Check that output populated correctly:

```{bash}
cat /Users/benzazycki/Documents/Research/ChickenEye/GRCg7b.110/seed.txt
```

Forge package:

```{r}
forgeBSgenomeDataPkg(seed_path, destdir = wd_path, replace = TRUE)
```

Build package with bash call:

```{bash}
cd /Users/benzazycki/Documents/Research/ChickenEye/GRCg7b.110
R CMD build BSgenome.Ggallus.ensembl.GRCg7b
```

Install newly forged/built package:

```{r}
package_path <- '/Users/benzazycki/Documents/Research/ChickenEye/GRCg7b.110/BSgenome.Ggallus.ensembl.GRCg7b_1.0.0.tar.gz'
install.packages(package_path, repos = NULL, type = "source")
```

## Generating Input Files

Next, we need to generate 4 files that will be used later in the pipeline:

-   1\. SeqInfo.csv: This stores important info about genomic sequences (names, lengths, etc.)

-   2\. Gene Transfer Format (.gtf): Describes gene/transcript features of the reference genome.

-   3\. W.txt: A list of W-chromosome linked genes.

-   4\. MT.txt: A list of mitochondria-linked genes.

### SeqInfo

Create data frame and output to file:

```{r}
seqinfo_df <- data.frame(seqnames = seqnames(seqinfo_data),
  length = seqlengths(seqinfo_data), isCircular = seqnames(seqinfo_data) %in% c("MT"))
csv_file <- "/Users/benzazycki/Documents/Research/ChickenEye/GRCg7b.110/GRCg7b.110.SeqInfo.csv"
write.csv(seqinfo_df, file = csv_file, row.names = FALSE, quote = FALSE)
```

### GTF

Simply download and unzip:

```{r}
gtf_url <- "https://ftp.ensembl.org/pub/release-110/gtf/gallus_gallus/Gallus_gallus.bGalGal1.mat.broiler.GRCg7b.110.gtf.gz"
gtf_zip_file <- "/Users/benzazycki/Documents/Research/ChickenEye/GRCg7b.110/GRCg7b.110.gtf.gz"
gtf_file <- "/Users/benzazycki/Documents/Research/ChickenEye/GRCg7b.110/GRCg7b.110.gtf"
download.file(gtf_url, destfile = gtf_zip_file)
R.utils::gunzip(gtf_zip_file, destname = gtf_file, remove=FALSE)
```

### W and MT

Import GTF file and extract features:

```{r}
gtf <- rtracklayer::import(gtf_file, format = "gtf")
genes <- gtf[gtf$type == "gene"]
```

Set preferred identifier and remove all NAs:

```{r}
get_preferred_id <- function(granges_obj) {
  ids <- ifelse(!is.na(granges_obj$gene_name),
               granges_obj$gene_name,
               granges_obj$gene_id)
  ids[!is.na(ids)] }
```

Extract Genes:

```{r}
genes_W <- genes[seqnames(genes) == "W"]
gene_ids_W <- get_preferred_id(genes_W)
genes_MT <- genes[seqnames(genes) == "MT"]
gene_ids_MT <- get_preferred_id(genes_MT)
```

Write output files:

```{r}
output_W <- "/Users/benzazycki/Documents/Research/ChickenEye/GRCg7b.110/GRCg7b.110.W.txt"
output_MT <- "/Users/benzazycki/Documents/Research/ChickenEye/GRCg7b.110/GRCg7b.110.MT.txt"
write.table(gene_ids_W, file = output_W,
            row.names = FALSE, col.names = FALSE, quote = FALSE)
write.table(gene_ids_MT, file = output_MT,
            row.names = FALSE, col.names = FALSE, quote = FALSE)
```

Lastly, list contents of working directory:

```{bash}
ls /Users/benzazycki/Documents/Research/ChickenEye/GRCg7b.110
```
