{"cells":[{"cell_type":"markdown","source":["# Chicken Eye: QA\n","Date: June 20 2025\n","\n","Author: Ben Zazycki\n","\n","Adapted from: Jared Tangeman\n","\n","Professor: Dr. Chun Liang\n","\n"],"metadata":{"id":"LXAfhioEhItW"}},{"cell_type":"markdown","source":["## Workspace Setup"],"metadata":{"id":"Okeb8vPEia-Y"}},{"cell_type":"code","execution_count":null,"metadata":{"collapsed":true,"id":"RqB69o1JJ_PE"},"outputs":[],"source":["from google.colab import drive\n","drive.mount('/content/drive')\n","!rm -rf /content/sample_data\n","!sudo apt-get install -y libgsl-dev\n","!sudo apt-get install -y libhdf5-dev\n","%load_ext rpy2.ipython\n","%R .libPaths(c('/content/drive/MyDrive/Bioinformatics/Colab_Lib/R', .libPaths()))\n","# ^ NOTE: change this based on your individual drive setup"]},{"cell_type":"markdown","metadata":{"id":"kRM2qnb_aaSJ"},"source":["Load relevant packages from library:"]},{"cell_type":"code","source":["%%R\n","library(Seurat)\n","library(Signac)\n","library(ggpubr)\n","library(ggplot2)\n","library(future)\n","library(DT)\n","library(gprofiler2)\n","library(scCustomize)\n","library(Matrix)\n","library(plotly)\n","library(ensembldb)\n","library(JASPAR2024)\n","library(DirichletMultinomial)\n","library(TFBSTools)\n","library(motifmatchr)\n","library(chromVAR)\n","library(ggforce)\n","library(GenomicRanges)\n","library(BSgenomeForge)\n","library(BSgenome)\n","library(biovizBase)\n","library(patchwork)\n","library(glmGamPoi)\n","library(presto)\n","library(GenomeInfoDb)\n","library(Biostrings)\n","library(rtracklayer)\n","library(BSgenome.Ggallus.ensembl.GRCg7b)"],"metadata":{"collapsed":true,"id":"yoseppMEvMlD"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Initial File Inputs"],"metadata":{"id":"JxWTpcwhUB6u"}},{"cell_type":"markdown","source":["Save path for file inputs:"],"metadata":{"id":"FHA3B34NTFcN"}},{"cell_type":"code","source":["%R input_path <- '/content/drive/MyDrive/Bioinformatics/Colab_Lib/Saved_Files/GRCg7b.110/'"],"metadata":{"collapsed":true,"id":"It31PPueTK1i"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Load in sequence info object:"],"metadata":{"id":"thr2XTF8UQoH"}},{"cell_type":"code","source":["%%R\n","seqInfo <- read.csv(paste0(input_path, 'GRCg7b.110.SeqInfo.csv'))\n","seqInfo <- Seqinfo(seqInfo$seqnames, seqlengths = seqInfo$length,\n","             isCircular = seqInfo$isCircular, genome = \"GRCg7b\")"],"metadata":{"id":"ilxcUp2KUSWi"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Create annotation from GTF file:"],"metadata":{"id":"8C8fi4sUUvFk"}},{"cell_type":"code","source":["%%R\n","edb <- EnsDb(ensDbFromGtf(paste0(input_path, 'GRCg7b.110.gtf'),\n","              organism='Gallus_gallus', genomeVersion='GRCg7b', version='110'))\n","annotation <- GetGRangesFromEnsDb(ensdb = edb, standard.chromosomes = FALSE)\n","genome(annotation) <- \"GRCg7b\""],"metadata":{"collapsed":true,"id":"8xPXlID0UX2n"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Introductory Explanation"],"metadata":{"id":"IdG5rSgLJvfX"}},{"cell_type":"markdown","source":["ATAC-seq Peaks\n","are genomic regions identified by peak-calling algorithms (like MACS2)\n","as having significant chromatin accessibility. They are stored as\n","a count matrix (cells x peaks), where each value indicates how many\n","ATAC-seq fragments overlap that peak in a given cell. Higher counts in a peak may suggest the chromatin region is more open. Peaks with high counts often overlap TB binding sites, enhancer/promoter, and regulatory elements."],"metadata":{"id":"J1MqDzp_J4a8"}},{"cell_type":"markdown","source":["These data represents discrete open chromatin regions (e.g., potential\n","enhancers, and promoters). They are used for identifying differential\n","accessible regions (DARs) and linking peaks to nearby genes (e.g. for\n","motif analysis)."],"metadata":{"id":"I3xgHfLMKHhE"}},{"cell_type":"markdown","source":["Fragment Files (fragpath_E4 = \"E4_atac_fragments.tsv.gz\") contains raw\n","fragment data listing all observed DNA fragments from ATAC-seq, with\n","genomic coordinates, cell barcode (linking fragments to cells) and UMI\n","(unique molecular identifier)."],"metadata":{"id":"zKZtCbxWKIHn"}},{"cell_type":"markdown","source":["In 10xGenomics, cell barcodes are unique sequences (typically 16 bp)\n","added to all molecules (RNA and ATAC) from the same cell during library\n","preparation, allowing cell-level pairing. All reads with the same\n","barcode are assumed to come from one cell. Cell barcodes are used to\n","distinguish real cells from background (e.g., empty droplets) via tools\n","like Cell Ranger."],"metadata":{"id":"L8OJr2e7KMPb"}},{"cell_type":"markdown","source":["UMIs (Unique Molecular Identifiers) are short random sequences\n","(typically 10–12 bp) added to individual RNA molecules during reverse\n","transcription. Not present in ATAC-seq data (UMIs are RNA-specific).\n","UMIs distinguish biological duplicates (true transcripts) from technical\n","duplicates (PCR artifacts). Unique UMI counts as one original molecule\n","(e.g., one mRNA transcript)."],"metadata":{"id":"knCM08buKNjD"}},{"cell_type":"markdown","source":["Fragment Files provide base-pair resolution of chromatin accessibility.\n","They can be Used for: (1) Calculating TSS enrichment and nucleosome\n","signal. (2) Visualizing insert size distributions (e.g., nucleosome-free\n","vs. nucleosome-bound fragments). (3) Recomputing peaks if needed."],"metadata":{"id":"MusBkOCTKP0l"}},{"cell_type":"markdown","source":["Key Differences: Feature ATAC Peaks Fragment Files Format Matrix (cells × peaks)\n","Tab-delimited (chrom, start, end, barcode) Resolution Regions (e.g.,\n","500bp windows) Single-base-pair (exact fragment boundaries) Content\n","Pre-defined open chromatin regions All observed DNA fragments Use Case\n","Peak-based analysis (DARs, motifs) QC, nucleosome positioning,\n","fine-scale analysis."],"metadata":{"id":"sRJYRE8kKRn-"}},{"cell_type":"markdown","source":["Peaks are used for quantitative analysis (e.g., \"How accessible is PeakX\n","in CellA vs. CellB?\"). Fragments are used for quality control (e.g., TSS\n","enrichment) and dynamic analyses (e.g., nucleosome positioning). Using\n","MACS2 or other tool, fragments data were used as the input to do Peak\n","Calling, obtaining Peak set as the output (counts_E4\\$Peaks)."],"metadata":{"id":"cVRnWxiBKXbm"}},{"cell_type":"markdown","source":["Most of this document resolves around Seurat objects which are initially split into separate objects for each developmental stage (E4, E5, E6, E7). Here is an explanation for the naming conventions present throughout this document:\n","*   `seu_E#_o`: original\n","*   `seu_E#`: filtered to only cells present in metadata\n","*   `seu_E#_n`: normalized\n","*   `seu_E#_f`: filtered\n","*   `seu_E#_ff`: filtered twice"],"metadata":{"id":"vllEdaoGvC1P"}},{"cell_type":"markdown","source":["## Main Data Reads"],"metadata":{"id":"rz0fRUoAW30v"}},{"cell_type":"code","source":["%R data_input_path <- '/content/drive/MyDrive/Bioinformatics/Lab_data/multiomics/'"],"metadata":{"collapsed":true,"id":"PPd0cQ2LW6yS"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Load in gene expression data from .h5 files:"],"metadata":{"id":"OVUBIcihXLfz"}},{"cell_type":"code","source":["%%R\n","counts_E4 <- Read10X_h5(filename = paste0(data_input_path, 'E4_filtered_feature_bc_matrix.h5'))\n","counts_E5 <- Read10X_h5(filename = paste0(data_input_path, 'E5_filtered_feature_bc_matrix.h5'))\n","counts_E6 <- Read10X_h5(filename = paste0(data_input_path, 'E6_filtered_feature_bc_matrix.h5'))\n","counts_E7 <- Read10X_h5(filename = paste0(data_input_path, 'E7_filtered_feature_bc_matrix.h5'))"],"metadata":{"collapsed":true,"id":"wBfhDVFtXTBT"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["The 10XGenomics .h5 files contain gene expression counts: Sparse matrix\n","where each row represents a gene and each column contains expressed\n","level (counts) for a given cell identified by a specific barcode. The\n",".h5 file also optionally contains ATAC peaks information, which is our\n","case here."],"metadata":{"id":"ZvDSL9hGJjH7"}},{"cell_type":"markdown","source":["Initialize original Seurat objects. The Seurat object stores (1)\n","counts: the raw gene expression matrix (2) assay='RNA': specifies that\n","this is RNA-seq data. The following will be automatically calculated: (1) nCount_RNA - Total reads per cell. (2) nFeature_RNA - Genes detected per cell  "],"metadata":{"id":"x24TSAtoX1EV"}},{"cell_type":"code","source":["%%R\n","seu_E4_o <- CreateSeuratObject(counts = counts_E4$`Gene Expression`, assay = \"RNA\")\n","seu_E5_o <- CreateSeuratObject(counts = counts_E5$`Gene Expression`, assay = \"RNA\")\n","seu_E6_o <- CreateSeuratObject(counts = counts_E6$`Gene Expression`, assay = \"RNA\")\n","seu_E7_o <- CreateSeuratObject(counts = counts_E7$`Gene Expression`, assay = \"RNA\")"],"metadata":{"collapsed":true,"id":"iuHMMzeyX9Tr"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Load in genotyping results from CSV files:"],"metadata":{"id":"sU9XRYSzYWgJ"}},{"cell_type":"code","source":["%%R\n","key_E4 <- read.csv(paste0(data_input_path, \"Genotypes_E4.csv\"), row.names = 1)\n","key_E5 <- read.csv(paste0(data_input_path, \"Genotypes_E5.csv\"), row.names = 1)\n","key_E6 <- read.csv(paste0(data_input_path, \"Genotypes_E6.csv\"), row.names = 1)\n","key_E7 <- read.csv(paste0(data_input_path, \"Genotypes_E7.csv\"), row.names = 1)"],"metadata":{"id":"TwkXi9WgYWLU"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["These files contain samples for each cell. Embryo has G, H and I whereas\n","Sex has Female and Male."],"metadata":{"id":"Y9V2IW_XKwgh"}},{"cell_type":"markdown","source":["## Initial Filtering, Editing, and Normalization"],"metadata":{"id":"L_vX0E2oZlyn"}},{"cell_type":"markdown","source":["Filter the seurat objects to keep only the cells present in the genotyping metadata:"],"metadata":{"id":"tNQygR0CZqZI"}},{"cell_type":"code","source":["%%R\n","seu_E4 <- subset(seu_E4_o, cells = rownames(key_E4))\n","seu_E5 <- subset(seu_E5_o, cells = rownames(key_E5))\n","seu_E6 <- subset(seu_E6_o, cells = rownames(key_E6))\n","seu_E7 <- subset(seu_E7_o, cells = rownames(key_E7))"],"metadata":{"id":"3FlCOZiPZoNf"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Reorder metadata of key_ objects (genotypes) so that it aligns with the seurat objects:"],"metadata":{"id":"NwNv3UAvZ2Nq"}},{"cell_type":"code","source":["%%R\n","key_E4 <- key_E4[rownames(seu_E4@meta.data), ]\n","key_E5 <- key_E5[rownames(seu_E5@meta.data), ]\n","key_E6 <- key_E6[rownames(seu_E6@meta.data), ]\n","key_E7 <- key_E7[rownames(seu_E7@meta.data), ]"],"metadata":{"id":"QmxG6e7YZ-LA"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Attach genotype metadata to each seurat object:"],"metadata":{"id":"xj4JdMVMaL-X"}},{"cell_type":"code","source":["%%R\n","seu_E4@meta.data <- cbind(seu_E4@meta.data, key_E4)\n","seu_E5@meta.data <- cbind(seu_E5@meta.data, key_E5)\n","seu_E6@meta.data <- cbind(seu_E6@meta.data, key_E6)\n","seu_E7@meta.data <- cbind(seu_E7@meta.data, key_E7)"],"metadata":{"id":"OLYO7WMAaPOr"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Assign an identity to each cell based on its developmental stage (E4-7):"],"metadata":{"id":"mcuADAW0aTGU"}},{"cell_type":"code","source":["%%R\n","seu_E4@meta.data$orig.ident <- seu_E4@meta.data$Stage\n","seu_E5@meta.data$orig.ident <- seu_E5@meta.data$Stage\n","seu_E6@meta.data$orig.ident <- seu_E6@meta.data$Stage\n","seu_E7@meta.data$orig.ident <- seu_E7@meta.data$Stage"],"metadata":{"id":"OQf8ROlFaX_m"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["NORMALIZATION: Apply Seurat's global-scale log-normalization method. Note: [Best Practices website](https://www.sc-best-practices.org/preprocessing_visualization/normalization.html) suggests Centered Log-Ratio (CLR)\n","transformation."],"metadata":{"id":"AX0XV5qRK9iX"}},{"cell_type":"code","source":["%%R\n","seu_E4_n <- NormalizeData(seu_E4)\n","seu_E5_n <- NormalizeData(seu_E5)\n","seu_E6_n <- NormalizeData(seu_E6)\n","seu_E7_n <- NormalizeData(seu_E7)"],"metadata":{"id":"4rHalui5arxv","collapsed":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Copy over metadata:"],"metadata":{"id":"D4mtilolauhg"}},{"cell_type":"code","source":["%%R\n","seu_E4_n@meta.data <- seu_E4@meta.data\n","seu_E5_n@meta.data <- seu_E5@meta.data\n","seu_E6_n@meta.data <- seu_E6@meta.data\n","seu_E7_n@meta.data <- seu_E7@meta.data"],"metadata":{"id":"qGjngRN1awEe"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Read in list of W-linked genes:"],"metadata":{"id":"beVMlI9SfLAq"}},{"cell_type":"code","source":["%R W <- readLines(\"/content/drive/MyDrive/Bioinformatics/Colab_Lib/Saved_Files/GRCg7b.110/GRCg7b.110.W.txt\")"],"metadata":{"id":"pXwcz-whf9kj","collapsed":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Read in list of mitochondrial genes:"],"metadata":{"id":"s1dmwBe2az3t"}},{"cell_type":"code","source":["%R MT <- readLines(\"/content/drive/MyDrive/Bioinformatics/Colab_Lib/Saved_Files/GRCg7b.110/GRCg7b.110.MT.txt\")"],"metadata":{"collapsed":true,"id":"QatYLbpjbY5u"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Calculate percentage of gene expression that is linked to mitochondria activity and add to metadata:"],"metadata":{"id":"vfGAiXN1bl0B"}},{"cell_type":"code","source":["%%R\n","seu_E4_n[[\"percent.mt\"]] <- PercentageFeatureSet(seu_E4_n, features = MT)\n","seu_E5_n[[\"percent.mt\"]] <- PercentageFeatureSet(seu_E5_n, features = MT)\n","seu_E6_n[[\"percent.mt\"]] <- PercentageFeatureSet(seu_E6_n, features = MT)\n","seu_E7_n[[\"percent.mt\"]] <- PercentageFeatureSet(seu_E7_n, features = MT)"],"metadata":{"id":"hhG231btblfJ"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Plot-Based Filtering"],"metadata":{"id":"3-jvdkYTb2Uy"}},{"cell_type":"markdown","source":["For each seurat object, I will plot (1) number of detected genes per cell, (2) number of RNA molecules per cell, and (3) percentage of mitochondrial expression per cell. These violin plots will then be used for QC filtering. Then, I will do the same plots again with the filtered objects."],"metadata":{"id":"EhToTBUhb9R1"}},{"cell_type":"markdown","source":["E4:"],"metadata":{"id":"LzF8bE0cg1IP"}},{"cell_type":"code","source":["%%R\n","VlnPlot(seu_E4_n, features = c(\"nFeature_RNA\",\"nCount_RNA\", \"percent.mt\"),\n","           ncol = 3, pt.size = 0)"],"metadata":{"id":"tmf-eLlwgfwm"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Perform filtering for E4:\n","* Number of unique genes: between 600 and 7000\n","* Number of RNA molecules per cell: Between 1000 and 15000\n","* Percentage of counts from mitochondrial genes: below 25%"],"metadata":{"id":"4Ra_IW2TiRjy"}},{"cell_type":"code","source":["%%R\n","seu_E4_f <- subset(seu_E4_n,\n","  subset = nFeature_RNA > 600 &\n","           nFeature_RNA < 7000 &\n","           percent.mt < 25 &\n","           nCount_RNA > 1000 &\n","           nCount_RNA < 15000)"],"metadata":{"id":"4xeNhkL9insw"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Post-filtering plot:"],"metadata":{"id":"pQYWNWCDi4JJ"}},{"cell_type":"code","source":["%%R\n","VlnPlot(seu_E4_f, features = c(\"nFeature_RNA\",\"nCount_RNA\", \"percent.mt\"),\n","           ncol = 3, pt.size = 0)"],"metadata":{"id":"A9TTLa2Ii3Ik"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["E5:"],"metadata":{"id":"nzf5IMLdg2Kc"}},{"cell_type":"code","source":["%%R\n","VlnPlot(seu_E5_n, features = c(\"nFeature_RNA\",\"nCount_RNA\", \"percent.mt\"),\n","           ncol = 3, pt.size = 0)"],"metadata":{"id":"a-h84en-ggOB"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Perform filtering for E5:\n","* Number of unique genes: between 600 and 6000\n","* Number of RNA molecules per cell: Between 1000 and 15000\n","* Percentage of counts from mitochondrial genes: below 20%"],"metadata":{"id":"vgAn8PGNjCuR"}},{"cell_type":"code","source":["%%R\n","seu_E5_f <- subset(seu_E5_n,\n","  subset = nFeature_RNA > 600 &\n","           nFeature_RNA < 6000 &\n","           percent.mt < 20 &\n","           nCount_RNA > 1000 &\n","           nCount_RNA < 15000)"],"metadata":{"id":"Hhs5sPsIjMEp"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Post-filtering plot:"],"metadata":{"id":"crHf1YxejWPc"}},{"cell_type":"code","source":["%%R\n","VlnPlot(seu_E5_f, features = c(\"nFeature_RNA\",\"nCount_RNA\", \"percent.mt\"),\n","           ncol = 3, pt.size = 0)"],"metadata":{"id":"iuj-vwXJjVQZ"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["E6:"],"metadata":{"id":"wce06hCzjEXr"}},{"cell_type":"code","source":["%%R\n","VlnPlot(seu_E6_n, features = c(\"nFeature_RNA\",\"nCount_RNA\", \"percent.mt\"),\n","           ncol = 3, pt.size = 0)"],"metadata":{"id":"PfPxQxzAggcD"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Perform filtering for E6:\n","* Number of unique genes: between 300 and 4500\n","* Number of RNA molecules per cell: Between 100 and 10000\n","* Percentage of counts from mitochondrial genes: below 15%"],"metadata":{"id":"OdDVAPJzjzOy"}},{"cell_type":"code","source":["%%R\n","seu_E6_f <- subset(seu_E6_n,\n","  subset = nFeature_RNA > 300 &\n","           nFeature_RNA < 4500 &\n","           percent.mt < 15 &\n","           nCount_RNA > 500 &\n","           nCount_RNA < 10000)"],"metadata":{"id":"oeNON6gZj6Ng"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Post-filtering plot:"],"metadata":{"id":"phWJvdrtj9Of"}},{"cell_type":"code","source":["%%R\n","VlnPlot(seu_E6_f, features = c(\"nFeature_RNA\",\"nCount_RNA\", \"percent.mt\"),\n","           ncol = 3, pt.size = 0)"],"metadata":{"id":"cftbd-y3j-Mt"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["E7:"],"metadata":{"id":"yImFSdHqg3od"}},{"cell_type":"code","source":["%%R\n","VlnPlot(seu_E7_n, features = c(\"nFeature_RNA\",\"nCount_RNA\", \"percent.mt\"),\n","           ncol = 3, pt.size = 0)"],"metadata":{"id":"ohPogz3Bgguj"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Perform filtering for E7:\n","* Number of unique genes: between 600 and 6500\n","* Number of RNA molecules per cell: Between 500 and 15000\n","* Percentage of counts from mitochondrial genes: below 20%"],"metadata":{"id":"EHuCuWhAkGce"}},{"cell_type":"code","source":["%%R\n","seu_E7_f <- subset(seu_E7_n,\n","  subset = nFeature_RNA > 600 &\n","           nFeature_RNA < 6500 &\n","           percent.mt < 20 &\n","           nCount_RNA > 500 &\n","           nCount_RNA < 15000)"],"metadata":{"id":"TW2UpXAjkPFr"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Post-filtering plot:"],"metadata":{"id":"N66hZ1CPkMkO"}},{"cell_type":"code","source":["%%R\n","VlnPlot(seu_E7_f, features = c(\"nFeature_RNA\",\"nCount_RNA\", \"percent.mt\"),\n","           ncol = 3, pt.size = 0)"],"metadata":{"id":"YUzznVd3kNWv"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Show merged violin plot to check overall results:"],"metadata":{"id":"7KpAqQBJkVyU"}},{"cell_type":"code","source":["%%R\n","seu_E4_f$stage <- \"E4\"\n","seu_E5_f$stage <- \"E5\"\n","seu_E6_f$stage <- \"E6\"\n","seu_E7_f$stage <- \"E7\"\n","seu_merged_f <- merge(seu_E4_f, y = list(seu_E5_f, seu_E6_f, seu_E7_f),\n","                    add.cell.ids = c(\"E4\", \"E5\", \"E6\", \"E7\"))\n","Idents(seu_merged_f) <- \"stage\"\n","VlnPlot(seu_merged_f,features = c(\"nFeature_RNA\", \"nCount_RNA\", \"percent.mt\"),\n","  pt.size = 0, ncol = 3)"],"metadata":{"id":"x3WbA4gwkXnx"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Adding on ATAC-seq Data"],"metadata":{"id":"dlLG3r56kvMp"}},{"cell_type":"markdown","source":["Now, we will add on ATAC-seq data to the already-existing Seurat objects. We'll do some basic QC and plotting."],"metadata":{"id":"xMfgzf7mkzf6"}},{"cell_type":"markdown","source":["Filter ATAC counts to match cell barcodes by subsetting the ATAC-seq peak counts to only include cells that are also present in the filtered RNA-based Seurat objects:"],"metadata":{"id":"V8nmo5tuk9k7"}},{"cell_type":"code","source":["%%R\n","atac_E4 <- counts_E4$Peaks[, colnames(counts_E4$Peaks) %in% colnames(seu_E4_f)]\n","atac_E5 <- counts_E5$Peaks[, colnames(counts_E5$Peaks) %in% colnames(seu_E5_f)]\n","atac_E6 <- counts_E6$Peaks[, colnames(counts_E6$Peaks) %in% colnames(seu_E6_f)]\n","atac_E7 <- counts_E7$Peaks[, colnames(counts_E7$Peaks) %in% colnames(seu_E7_f)]"],"metadata":{"id":"dhzUvGDUkyex"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Create objects that hold the path to fragment files:"],"metadata":{"id":"2tasmZTrlJj1"}},{"cell_type":"code","source":["%%R\n","fragpath_E4 <- paste0(data_input_path, \"E4_atac_fragments.tsv.gz\")\n","fragpath_E5 <- paste0(data_input_path, \"E5_atac_fragments.tsv.gz\")\n","fragpath_E6 <- paste0(data_input_path, \"E6_atac_fragments.tsv.gz\")\n","fragpath_E7 <- paste0(data_input_path, \"E7_atac_fragments.tsv.gz\")"],"metadata":{"id":"2yWhDZ-5lMkA"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Add ATAC assay to each object:"],"metadata":{"id":"V0TiaOn6lDXM"}},{"cell_type":"code","source":["%%R\n","seu_E4_f[[\"ATAC\"]] <- CreateChromatinAssay(counts = atac_E4,\n","                                         sep = c(\":\", \"-\"),\n","                                         fragments = fragpath_E4,\n","                                         annotation = annotation)\n","\n","seu_E5_f[[\"ATAC\"]] <- CreateChromatinAssay(counts = atac_E5,\n","                                         sep = c(\":\", \"-\"),\n","                                         fragments = fragpath_E5,\n","                                         annotation = annotation)\n","\n","seu_E6_f[[\"ATAC\"]] <- CreateChromatinAssay(counts = atac_E6,\n","                                         sep = c(\":\", \"-\"),\n","                                         fragments = fragpath_E6,\n","                                         annotation = annotation)\n","\n","seu_E7_f[[\"ATAC\"]] <- CreateChromatinAssay(counts = atac_E7,\n","                                         sep = c(\":\", \"-\"),\n","                                         fragments = fragpath_E7,\n","                                         annotation = annotation)"],"metadata":{"collapsed":true,"id":"DGfY3qz9lEl4"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Compute Nucleosome Signal and TSS Enrichment (QC metrics):"],"metadata":{"id":"A9vol9SUmJon"}},{"cell_type":"code","source":["%%R\n","DefaultAssay(seu_E4_f) <- \"ATAC\"\n","DefaultAssay(seu_E5_f) <- \"ATAC\"\n","DefaultAssay(seu_E6_f) <- \"ATAC\"\n","DefaultAssay(seu_E7_f) <- \"ATAC\"\n","\n","seu_E4_f <- NucleosomeSignal(seu_E4_f)\n","seu_E5_f <- NucleosomeSignal(seu_E5_f)\n","seu_E6_f <- NucleosomeSignal(seu_E6_f)\n","seu_E7_f <- NucleosomeSignal(seu_E7_f)\n","\n","seu_E4_f <- TSSEnrichment(seu_E4_f)\n","seu_E5_f <- TSSEnrichment(seu_E5_f)\n","seu_E6_f <- TSSEnrichment(seu_E6_f)\n","seu_E7_f <- TSSEnrichment(seu_E7_f)"],"metadata":{"id":"8I9qOw5EmUnt","collapsed":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Specify clean levels:"],"metadata":{"id":"XOfXUJqLmh58"}},{"cell_type":"code","source":["%%R\n","seu_E4_f <- subset(seu_E4_f, subset = orig.ident == \"E4\")\n","seu_E5_f <- subset(seu_E5_f, subset = orig.ident == \"E5\")\n","seu_E6_f <- subset(seu_E6_f, subset = orig.ident == \"E6\")\n","seu_E7_f <- subset(seu_E7_f, subset = orig.ident == \"E7\")"],"metadata":{"id":"zfbDVrObmjfw"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## ATAC-seq Plotting"],"metadata":{"id":"wh6qZ2GFmpEi"}},{"cell_type":"markdown","source":["Each stage will get 3 more plots:\n","\n","*   DensityScatter to visualize relationship between ATAC reads and TSS enrichment\n","*   Violin plot with RNA counts and ATAC metrics\n","* Histogram to visualize fragment lengths\n","\n","Then, more QC will be perfomed based on the results.\n","\n"],"metadata":{"id":"Enxj5p7fmugu"}},{"cell_type":"markdown","source":["The fragment histograms show the size distribution of DNA fragments after\n","Tn5 transposase digestion, which is critical for assessing ATAC-seq\n","library quality."],"metadata":{"id":"ebjBLmkILeul"}},{"cell_type":"markdown","source":["~200 bp: Mononucleosome fragments (DNA wrapped around 1 nucleosome +\n","linker).\n","\n","~400 bp: Dinucleosome fragments (2 nucleosomes + linker).\n","\n","<100 bp: Open chromatin (Tn5 cut sites without nucleosomes).\n","\n","Sharp peak <100 bp (open chromatin). Smaller peaks at ~200 bp and\n","~400 bp (nucleosome-associated fragments)."],"metadata":{"id":"BFDSCJb1LlWj"}},{"cell_type":"markdown","source":["E4:"],"metadata":{"id":"D-ymqYSCnCyZ"}},{"cell_type":"code","source":["%%R\n","DensityScatter(seu_E4_f, x = 'nCount_ATAC', y = 'TSS.enrichment',\n","               log_x = TRUE, quantiles = TRUE)"],"metadata":{"id":"TG8eigVfmrcl"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","VlnPlot(object = seu_E4_f, features = c(\"nCount_RNA\", \"nCount_ATAC\",\n","                                  \"TSS.enrichment\", \"nucleosome_signal\"),\n","        ncol = 4, pt.size = 0, group.by = \"orig.ident\")"],"metadata":{"id":"Urk44-SHnJVn"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","FragmentHistogram(object = seu_E4_f, region = '1-1-100000000',\n","                  group.by = \"orig.ident\", assay = \"ATAC\")"],"metadata":{"id":"OLTBp0AxnLkq"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Perform QC:\n","\n","\n","*   ATAC reads between 500 and 50000\n","*   Nucleosome signal less than 1\n","*   TSS enrichment betwen 1 and 10\n","\n"],"metadata":{"id":"VPyYkxzunOM5"}},{"cell_type":"code","source":["%%R\n","seu_E4_ff <- subset(x = seu_E4_f,\n","                subset = nCount_ATAC < 50000 & nCount_ATAC > 500 &\n","                nucleosome_signal < 1 &\n","                TSS.enrichment > 1 & TSS.enrichment < 10)"],"metadata":{"id":"jt1S7VQ5nZ7K"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["E5:"],"metadata":{"id":"Pwl3EuWFnfpb"}},{"cell_type":"code","source":["%%R\n","DensityScatter(seu_E5_f, x = 'nCount_ATAC', y = 'TSS.enrichment',\n","               log_x = TRUE, quantiles = TRUE)"],"metadata":{"id":"WaALZlW6nfpc"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","VlnPlot(object = seu_E5_f, features = c(\"nCount_RNA\", \"nCount_ATAC\",\n","                                  \"TSS.enrichment\", \"nucleosome_signal\"),\n","        ncol = 4, pt.size = 0, group.by = \"orig.ident\")"],"metadata":{"id":"hMmMb-Fmnfpc"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","FragmentHistogram(object = seu_E5_f, region = '1-1-100000000',\n","                  group.by = \"orig.ident\", assay = \"ATAC\")"],"metadata":{"id":"ut1hmSXAnfpc"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Perform QC:\n","\n","\n","*   ATAC reads between 500 and 50000\n","*   Nucleosome signal less than 1\n","*   TSS enrichment betwen 1 and 10\n","\n"],"metadata":{"id":"2kjH8pCxnfpc"}},{"cell_type":"code","source":["%%R\n","seu_E5_ff <- subset(x = seu_E5_f,\n","                subset = nCount_ATAC < 50000 & nCount_ATAC > 500 &\n","                nucleosome_signal < 1 &\n","                TSS.enrichment > 1 & TSS.enrichment < 10)"],"metadata":{"id":"aKRZ5N2bnfpc"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["E6:"],"metadata":{"id":"b9WeJPsAoBvq"}},{"cell_type":"code","source":["%%R\n","DensityScatter(seu_E6_f, x = 'nCount_ATAC', y = 'TSS.enrichment',\n","               log_x = TRUE, quantiles = TRUE)"],"metadata":{"id":"Ytx5yUkZoBvr"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","VlnPlot(object = seu_E6_f, features = c(\"nCount_RNA\", \"nCount_ATAC\",\n","                                  \"TSS.enrichment\", \"nucleosome_signal\"),\n","        ncol = 4, pt.size = 0, group.by = \"orig.ident\")"],"metadata":{"id":"JMfP38TZoBvr"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","FragmentHistogram(object = seu_E6_f, region = '1-1-100000000',\n","                  group.by = \"orig.ident\", assay = \"ATAC\")"],"metadata":{"id":"w6qQX1zZoBvr"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Perform QC:\n","\n","\n","*   ATAC reads between 354 and 20000\n","*   Nucleosome signal less than 1\n","*   TSS enrichment betwen 1 and 10\n","\n"],"metadata":{"id":"so4dZR57oBvr"}},{"cell_type":"code","source":["%%R\n","seu_E6_ff <- subset(x = seu_E6_f,\n","                subset = nCount_ATAC < 20000 & nCount_ATAC > 354 &\n","                nucleosome_signal < 1 &\n","                TSS.enrichment > 1 & TSS.enrichment < 10)"],"metadata":{"id":"2bTu4BqWoBvr"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["E7:"],"metadata":{"id":"LuHtcCJ1odwI"}},{"cell_type":"code","source":["%%R\n","DensityScatter(seu_E7_f, x = 'nCount_ATAC', y = 'TSS.enrichment',\n","               log_x = TRUE, quantiles = TRUE)"],"metadata":{"id":"kkNCvdPmodwJ"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","VlnPlot(object = seu_E7_f, features = c(\"nCount_RNA\", \"nCount_ATAC\",\n","                                  \"TSS.enrichment\", \"nucleosome_signal\"),\n","        ncol = 4, pt.size = 0, group.by = \"orig.ident\")"],"metadata":{"id":"b5E4gubTodwJ"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","FragmentHistogram(object = seu_E7_f, region = '1-1-100000000',\n","                  group.by = \"orig.ident\", assay = \"ATAC\")"],"metadata":{"id":"8pbZzinOodwJ"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Perform QC:\n","\n","\n","*   ATAC reads between 484 and 30000\n","*   Nucleosome signal less than 1\n","*   TSS enrichment betwen 1 and 10\n","\n"],"metadata":{"id":"U-VecdKeodwJ"}},{"cell_type":"code","source":["%%R\n","seu_E7_ff <- subset(x = seu_E7_f,\n","                subset = nCount_ATAC < 30000 & nCount_ATAC > 484 &\n","                nucleosome_signal < 1 &\n","                TSS.enrichment > 1 & TSS.enrichment < 10)"],"metadata":{"id":"gGNu2Du4odwJ"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Further ATAC processing"],"metadata":{"id":"G_VAves6KuFq"}},{"cell_type":"markdown","source":["Extract the genomic ranges of peak regions from the seurat objects:"],"metadata":{"id":"_6UBPqsTcBQW"}},{"cell_type":"code","source":["%%R\n","peak_E4 <- seu_E4_ff@assays$ATAC@ranges\n","peak_E5 <- seu_E5_ff@assays$ATAC@ranges\n","peak_E6 <- seu_E6_ff@assays$ATAC@ranges\n","peak_E7 <- seu_E7_ff@assays$ATAC@ranges"],"metadata":{"id":"EQORCMiXKwGl"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Combine all peak regions into one unified set: (Note:\n","GenomicRanges::reduce() removes duplicates)"],"metadata":{"id":"SEUhBn6hcU3_"}},{"cell_type":"code","source":["%R combined.peaks <- reduce(x = c(peak_E4, peak_E5, peak_E6, peak_E7))"],"metadata":{"collapsed":true,"id":"PBOfdw0PcVm2"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Read in fragment files to make objects:\n"],"metadata":{"id":"ecboz5c8caAV"}},{"cell_type":"code","source":["%%R\n","frags.seu_E4_ff <- CreateFragmentObject(\n","  path = \"/content/drive/MyDrive/Bioinformatics/Lab_data/multiomics/E4_atac_fragments.tsv.gz\",\n","  cells = colnames(seu_E4_ff))\n","frags.seu_E5_ff <- CreateFragmentObject(\n","  path = \"/content/drive/MyDrive/Bioinformatics/Lab_data/multiomics/E5_atac_fragments.tsv.gz\",\n","  cells = colnames(seu_E5_ff))\n","frags.seu_E6_ff <- CreateFragmentObject(\n","  path = \"/content/drive/MyDrive/Bioinformatics/Lab_data/multiomics/E6_atac_fragments.tsv.gz\",\n","  cells = colnames(seu_E6_ff))\n","frags.seu_E7_ff <- CreateFragmentObject(\n","  path = \"/content/drive/MyDrive/Bioinformatics/Lab_data/multiomics/E7_atac_fragments.tsv.gz\",\n","  cells = colnames(seu_E7_ff))"],"metadata":{"collapsed":true,"id":"gnVkgSyEcbpT"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Create FeatureMatrix objects: (Note: These map the number of reads from\n","a given cell that overlap with a given genomic peak from\n","'combined.peaks'.)"],"metadata":{"id":"HXf4A-Nmc4F2"}},{"cell_type":"code","source":["%%R\n","seu_E4_ff.counts <- FeatureMatrix(fragments = frags.seu_E4_ff,\n","                   features = combined.peaks, cells = colnames(seu_E4_ff))\n","seu_E5_ff.counts <- FeatureMatrix(fragments = frags.seu_E5_ff,\n","                   features = combined.peaks, cells = colnames(seu_E5_ff))\n","seu_E6_ff.counts <- FeatureMatrix(fragments = frags.seu_E6_ff,\n","                   features = combined.peaks, cells = colnames(seu_E6_ff))\n","seu_E7_ff.counts <- FeatureMatrix(fragments = frags.seu_E7_ff,\n","                   features = combined.peaks, cells = colnames(seu_E7_ff))"],"metadata":{"collapsed":true,"id":"zNcT7J0Bc4be"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Rebuild the ATAC assays using the new combined peak set and updated\n","fragment data:"],"metadata":{"id":"LcMuMpQTdn14"}},{"cell_type":"code","source":["%%R\n","seu_E4_ff[[\"ATAC\"]]  <- CreateChromatinAssay(seu_E4_ff.counts,\n","                        fragments = frags.seu_E4_ff,  annotation = annotation)\n","seu_E5_ff[[\"ATAC\"]]  <- CreateChromatinAssay(seu_E5_ff.counts,\n","                        fragments = frags.seu_E5_ff,  annotation = annotation)\n","seu_E6_ff[[\"ATAC\"]]  <- CreateChromatinAssay(seu_E6_ff.counts,\n","                        fragments = frags.seu_E6_ff,  annotation = annotation)\n","seu_E7_ff[[\"ATAC\"]]  <- CreateChromatinAssay(seu_E7_ff.counts,\n","                        fragments = frags.seu_E7_ff,  annotation = annotation)"],"metadata":{"collapsed":true,"id":"CZBpBotSdpPE"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Final (Merged) Processing"],"metadata":{"id":"VjdI3AiVfSXu"}},{"cell_type":"markdown","source":["Merge the Seurat objects for final processing:"],"metadata":{"id":"M4GNYnWbeN2X"}},{"cell_type":"code","source":["%R seu_merged <- merge(x = seu_E4_ff, y = list(seu_E5_ff, seu_E6_ff, seu_E7_ff))"],"metadata":{"collapsed":true,"id":"KPt8-mUKeOLc"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Filter to only useful metadata columns:\n"],"metadata":{"id":"TivP9YkmevUx"}},{"cell_type":"code","source":["%%R\n","DefaultAssay(seu_merged) <- \"RNA\"\n","seu_merged@meta.data <- subset(seu_merged@meta.data,\n","    select=c(orig.ident, Embryo, Sex, nCount_RNA, nFeature_RNA, nCount_ATAC, nFeature_ATAC))"],"metadata":{"id":"HpMPgcmQewsy"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Join data across layers and normalize:"],"metadata":{"id":"ZQmsz_Kye10Z"}},{"cell_type":"code","source":["%%R\n","seu_merged <- JoinLayers(seu_merged)\n","seu_merged <- NormalizeData(seu_merged)"],"metadata":{"collapsed":true,"id":"sFOHpa2we3TL"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Calculate the percentage of RNA reads mapping to mitochondrial and W-linked genes:\n"],"metadata":{"id":"JC60LP4SiIhq"}},{"cell_type":"code","source":["%%R\n","seu_merged[[\"percent.mt\"]] <- PercentageFeatureSet(seu_merged, features = MT)\n","seu_merged[[\"percent.w\"]] <- PercentageFeatureSet(seu_merged, features = W)"],"metadata":{"id":"_pAxyCCQiKc0"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Calculate Nuclesome Signal Score and TSS Enrichment Score for each cell:"],"metadata":{"id":"x3b9w3sBiSIP"}},{"cell_type":"code","source":["%%R\n","DefaultAssay(seu_merged) <- \"ATAC\"\n","seu_merged <- NucleosomeSignal(seu_merged)\n","seu_merged <- TSSEnrichment(seu_merged)"],"metadata":{"collapsed":true,"id":"zaZz5rdaiZA1"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Visualize with Count-Enrichment scatter plot:"],"metadata":{"id":"GeBVL7EnigZw"}},{"cell_type":"code","source":["%%R\n","DensityScatter(seu_merged, x = 'nCount_ATAC', y = 'TSS.enrichment',\n","               log_x = TRUE, quantiles = TRUE)"],"metadata":{"id":"ny6YV9W9iiGo"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Visualize violin plot with 6 values:"],"metadata":{"id":"pvuh6xoaikE6"}},{"cell_type":"code","source":["%%R\n","VlnPlot(object = seu_merged,\n","        features = c(\"nFeature_RNA\",\n","                     \"nCount_RNA\",\n","                     \"percent.mt\",\n","                     \"nCount_ATAC\",\n","                     \"TSS.enrichment\",\n","                     \"nucleosome_signal\"),\n","        ncol = 3, pt.size = 0, group.by = \"orig.ident\")"],"metadata":{"id":"I5eFtYyNikZe"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Next, I will calculate cell cycle phase score AND cell cycle difference score. Cell cycle phase scores are calculated using predefined S phase and G2/M phase marker genes. Cell cycle difference scores are numeric values representing each cell's relative bias towards S phase vs. G2/M."],"metadata":{"id":"gWPO3EsspHXm"}},{"cell_type":"code","source":["%%R\n","DefaultAssay(seu_merged) <- \"RNA\"\n","seu_merged <- CellCycleScoring(seu_merged, s.features = cc.genes$s.genes,\n","              g2m.features = cc.genes$g2m.genes,set.ident = FALSE)\n","seu_merged$CC.Difference <- seu_merged$S.Score - seu_merged$G2M.Score"],"metadata":{"collapsed":true,"id":"4AqtXWPSo9qx"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Perform SCTransform normalization of gene expression:"],"metadata":{"id":"GIC5bivtFAY9"}},{"cell_type":"code","source":["%%R\n","seu_merged <- SCTransform(seu_merged, assay = \"RNA\",\n","                vars.to.regress = c(\"CC.Difference\", \"percent.mt\", \"percent.w\"))\n","DefaultAssay(seu_merged) <- \"SCT\""],"metadata":{"collapsed":true,"id":"Dx-03mYrFA4s"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Saving Processed Data"],"metadata":{"id":"uDd2nKEANXIP"}},{"cell_type":"markdown","source":["Now, I will demonstrate saving the processed data from this session in order to be used in the next notebook in the pipeline. I will save the merged Seurat object, annotation object, sequence info object, and the R session info."],"metadata":{"id":"OIb_6AXrG80_"}},{"cell_type":"code","source":["%%R\n","output_dir <- '/content/drive/MyDrive/Bioinformatics/Colab_Lib/Saved_Files/GRCg7b.110/Data_Outputs'\n","saveRDS(seu_merged, file = file.path(output_dir, \"seu_merged_processed.rds\"))\n","saveRDS(annotation, file = file.path(output_dir, \"annotation.rds\"))\n","saveRDS(seqInfo, file = file.path(output_dir, \"seqInfo.rds\"))\n","writeLines(capture.output(sessionInfo()), file.path(output_dir, \"session_info.txt\"))"],"metadata":{"id":"q_7jztiaG8nb"},"execution_count":null,"outputs":[]}],"metadata":{"colab":{"provenance":[],"machine_shape":"hm","authorship_tag":"ABX9TyOBolzpWZzuSGJ22UUJRaWU"},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"name":"python"}},"nbformat":4,"nbformat_minor":0}