{"cells":[{"cell_type":"markdown","source":["# Chicken Eye: Setup\n","Date: June 20 2025\n","\n","Author: Ben Zazycki\n","\n","Adapted from: Jared Tangeman\n","\n","Professor: Dr. Chun Liang\n"],"metadata":{"id":"LXAfhioEhItW"}},{"cell_type":"markdown","source":["## Workspace Setup"],"metadata":{"id":"yJICw6C7guRN"}},{"cell_type":"code","execution_count":null,"metadata":{"collapsed":true,"id":"RqB69o1JJ_PE"},"outputs":[],"source":["from google.colab import drive\n","drive.mount('/content/drive')\n","!rm -rf /content/sample_data\n","!sudo apt-get install -y libgsl-dev\n","!sudo apt-get install -y libhdf5-dev\n","%load_ext rpy2.ipython\n","%R .libPaths(c('/content/drive/MyDrive/Bioinformatics/Colab_Lib/R', .libPaths()))\n","# ^ NOTE: change this based on your individual drive setup"]},{"cell_type":"markdown","metadata":{"id":"W7_HQfYANW8m"},"source":["## Installing Packages"]},{"cell_type":"markdown","metadata":{"id":"kmdwEIqNNbfP"},"source":["Note 1: Only run the cells to install the packages you haven't installed yet."]},{"cell_type":"markdown","source":["Note 2: Check the outputs for the install lines. If it's not being installed into your specified libPath from above, ensure that you ran that line. If it's still not working, you can specify `'lib = /path/to/lib'` after the package name. The first line below is an example of this."],"metadata":{"id":"4Ic8vnQTCqHF"}},{"cell_type":"markdown","source":["Note 3: There are two methods for installing R packages on colab. First, you can simply use R cell magic with `%R` or `%%R`. However, you can also use the command line call: `!R -e \"insert code here\"`. This is more reliable for installing some packages. It's important to note that each new call of `!R -e` will initialize a new R session. Installing packages with BiocManager requires having BiocManager installed and loaded into the session, so it is not possible with this method. If one method isn't working to install a specific package for you, try the other method."],"metadata":{"id":"8kC9dmukFKFK"}},{"cell_type":"code","source":["!R -e \"install.packages('remotes', lib = '/content/drive/MyDrive/Bioinformatics/Colab_Lib/R')\""],"metadata":{"id":"Hd5rMS8hLU4j"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["!R -e \"install.packages('devtools')\""],"metadata":{"id":"-IAKqoPxLVod"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%R remotes::install_github('satijalab/seurat-data')"],"metadata":{"id":"kMcWEdj5GO8O"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%R install.packages(\"hdf5r\")"],"metadata":{"id":"PBJRK4J1kKf3"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%R remotes::install_github('10XGenomics/loupeR')\""],"metadata":{"id":"HVLoEM-Tj4H0"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%R devtools::install_github(\"immunogenomics/presto\")"],"metadata":{"id":"h0TM-P7-GPTZ"},"execution_count":null,"outputs":[]},{"cell_type":"code","execution_count":null,"metadata":{"id":"I1ev8Ln7Ottc"},"outputs":[],"source":["!R -e \"install.packages('BiocManager')\""]},{"cell_type":"code","execution_count":null,"metadata":{"id":"yDcS5CaxSmbY"},"outputs":[],"source":["!R -e \"install.packages('reticulate')\""]},{"cell_type":"code","execution_count":null,"metadata":{"id":"s_UamSTDNhIt"},"outputs":[],"source":["!R -e \"install.packages('Seurat')\""]},{"cell_type":"code","execution_count":null,"metadata":{"id":"xe7z_V1lP0Iq"},"outputs":[],"source":["!R -e \"install.packages('Signac')\""]},{"cell_type":"code","execution_count":null,"metadata":{"id":"RZncBnzrP6ny"},"outputs":[],"source":["!R -e \"install.packages('ggpubr')\""]},{"cell_type":"code","execution_count":null,"metadata":{"id":"AQJGEgggP8jf"},"outputs":[],"source":["!R -e \"install.packages('ggplot2')\""]},{"cell_type":"code","execution_count":null,"metadata":{"id":"MlGElsPJP-K7"},"outputs":[],"source":["!R -e \"install.packages('future')\""]},{"cell_type":"code","execution_count":null,"metadata":{"id":"Hyaect0BP_cW"},"outputs":[],"source":["!R -e \"install.packages('DT')\""]},{"cell_type":"code","execution_count":null,"metadata":{"id":"Ie56WcC0QCXn"},"outputs":[],"source":["!R -e \"install.packages('gprofiler2')\""]},{"cell_type":"code","execution_count":null,"metadata":{"id":"rk_swRmgQDlO"},"outputs":[],"source":["!R -e \"install.packages('Matrix')\""]},{"cell_type":"code","execution_count":null,"metadata":{"id":"2g2smzB9QErB"},"outputs":[],"source":["!R -e \"install.packages('plotly')\""]},{"cell_type":"code","execution_count":null,"metadata":{"id":"VroKff3aQFt9"},"outputs":[],"source":["!R -e \"install.packages('ggforce')\""]},{"cell_type":"code","execution_count":null,"metadata":{"id":"Jb7DvDWJQG9s"},"outputs":[],"source":["!R -e \"install.packages('patchwork')\""]},{"cell_type":"code","execution_count":null,"metadata":{"id":"sxgzjo-_QIGC"},"outputs":[],"source":["!R -e \"install.packages('presto')\""]},{"cell_type":"code","execution_count":null,"metadata":{"id":"bXl5ZzleQJA6"},"outputs":[],"source":["!R -e \"install.packages('scCustomize')\""]},{"cell_type":"code","execution_count":null,"metadata":{"id":"z52nmGyRSH_I","collapsed":true},"outputs":[],"source":["%R BiocManager::install('glmGamPoi')"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"bRZTSZ9qQLvr"},"outputs":[],"source":["%R BiocManager::install('ensembldb')"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"qiJN4RJhQOAM"},"outputs":[],"source":["%R BiocManager::install('JASPAR2024')"]},{"cell_type":"code","execution_count":null,"metadata":{"collapsed":true,"id":"qVYsuQdJQPVW"},"outputs":[],"source":["%R BiocManager::install('TFBSTools')"]},{"cell_type":"code","execution_count":null,"metadata":{"collapsed":true,"id":"-SNaRaytQQg4"},"outputs":[],"source":["%R BiocManager::install('motifmatchr')"]},{"cell_type":"code","execution_count":null,"metadata":{"collapsed":true,"id":"K6W2ZmfnQSmB"},"outputs":[],"source":["%R BiocManager::install('chromVAR')"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"8TTe3-2FQT0W"},"outputs":[],"source":["%R BiocManager::install('GenomicRanges')"]},{"cell_type":"code","execution_count":null,"metadata":{"collapsed":true,"id":"LkIU4QQkQVMX"},"outputs":[],"source":["%R BiocManager::install('BSgenomeForge')"]},{"cell_type":"code","execution_count":null,"metadata":{"collapsed":true,"id":"prQeNsbwQWVe"},"outputs":[],"source":["%R BiocManager::install('biovizBase')"]},{"cell_type":"code","source":["%R BiocManager::install('DirichletMultinomial')"],"metadata":{"collapsed":true,"id":"iFlbhdSzY5fS"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%R BiocManager::install('GenomeInfoDb')"],"metadata":{"id":"sPR9plD-jKc9"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%R BiocManager::install('BSgenome')"],"metadata":{"id":"_rWLxuANja4i"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%R BiocManager::install('Biostrings')"],"metadata":{"id":"y5X_cG1Xjkcd"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%R BiocManager::install('rtracklayer')"],"metadata":{"id":"tNBu4OcNjmF7"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%R install.packages('R.utils')"],"metadata":{"id":"1oXWVju5VTXr"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"kRM2qnb_aaSJ"},"source":["## Loading from Libraries"]},{"cell_type":"markdown","source":["Load the packages neccessary for this setup document:"],"metadata":{"id":"DWCvsvXLVi_J"}},{"cell_type":"code","source":["%%R\n","library(R.utils)\n","library(Biostrings)\n","library(rtracklayer)\n","library(GenomeInfoDb)\n","library(BSgenome)\n","library(BSgenomeForge)"],"metadata":{"collapsed":true,"id":"yoseppMEvMlD"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Forging Custom Package"],"metadata":{"id":"pNWY51CkiTOV"}},{"cell_type":"markdown","source":["In this section, I will demonstrate creating a custome BSgenomeForge package for the relevant genome. This section involves a lot of work with filepaths. Be sure you change everything to match the setup on your local machine."],"metadata":{"id":"B5lfG0GHKLie"}},{"cell_type":"markdown","source":["First, make a directory to work in:"],"metadata":{"id":"3GxgiVx1KL46"}},{"cell_type":"code","source":["!mkdir /content/drive/MyDrive/Bioinformatics/Colab_Lib/Saved_Files/GRCg7b.110"],"metadata":{"id":"T6nz5FlzXlGX"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Download and unzip genome sequence file from ENSEMBL website"],"metadata":{"id":"HT7Q_yR6kUKI"}},{"cell_type":"code","source":["%%R\n","ensembl_url <- \"ftp://ftp.ensembl.org/pub/release-110/fasta/gallus_gallus/dna/Gallus_gallus.bGalGal1.mat.broiler.GRCg7b.dna.toplevel.fa.gz\"\n","download.file(ensembl_url, destfile = \"/content/drive/MyDrive/Bioinformatics/Colab_Lib/Saved_Files/GRCg7b.110/Gallus_gallus.GRCg7b.fa.gz\")\n","R.utils::gunzip(\"/content/drive/MyDrive/Bioinformatics/Colab_Lib/Saved_Files/GRCg7b.110/Gallus_gallus.GRCg7b.fa.gz\", remove = FALSE)"],"metadata":{"collapsed":true,"id":"Ig5-tStqiWmd"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Read FASTA as DNAStringSet"],"metadata":{"id":"f7Pl4A3fo8zG"}},{"cell_type":"code","source":["%%R\n","fasta_path <- \"/content/drive/MyDrive/Bioinformatics/Colab_Lib/Saved_Files/GRCg7b.110/Gallus_gallus.GRCg7b.fa\"\n","genome_seq <- readDNAStringSet(fasta_path)"],"metadata":{"id":"EpYZgN03omq9"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Export to .2bit format"],"metadata":{"id":"CBTOqffhpPZB"}},{"cell_type":"code","source":["%%R\n","twobit_file <- \"/content/drive/MyDrive/Bioinformatics/Colab_Lib/Saved_Files/GRCg7b.110/Gallus_gallus.GRCg7b.2bit\"\n","export(genome_seq, twobit_file, format = \"2bit\")"],"metadata":{"id":"ZCfnZOijpQHS"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Get sequence lengths from the 2bit file"],"metadata":{"id":"mSSCEDmYp0hz"}},{"cell_type":"code","source":["%%R\n","two_bit <- TwoBitFile(twobit_file)\n","seqinfo_data <- seqinfo(two_bit)"],"metadata":{"id":"_bZi34qep1Ij"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Create Seqinfo object with explicit genome name, save for BSgenome"],"metadata":{"id":"CniIp-xvqDgY"}},{"cell_type":"code","source":["%%R\n","seqinfo_obj <- Seqinfo(\n","  seqnames = seqnames(seqinfo_data),\n","  seqlengths = seqlengths(seqinfo_data),\n","  genome = \"GRCg7b\")\n","saveRDS(seqinfo_obj, \"/content/drive/MyDrive/Bioinformatics/Colab_Lib/Saved_Files/GRCg7b.110/seqinfo.rds\")"],"metadata":{"id":"EargnXVFqEeq"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Prepare seed file"],"metadata":{"id":"9edb1Hpxsg8_"}},{"cell_type":"code","source":["%%R\n","seqnames_data <- seqnames(seqinfo_data)\n","seed_content <- sprintf(\n","\"Package: BSgenome.Ggallus.ensembl.GRCg7b\n","Title: Gallus gallus (GRCg7b) from ENSEMBL\n","Description: Full genome from ENSEMBL release 110\n","Version: 1.0.0\n","organism: Gallus gallus\n","common_name: Chicken\n","provider: ENSEMBL\n","provider_version: GRCg7b\n","release_date: 2023\n","source_url: ftp://ftp.ensembl.org/pub/release-110/fasta/gallus_gallus/dna/\n","seqs_srcdir: /content/drive/MyDrive/Bioinformatics/Colab_Lib/Saved_Files/GRCg7b.110\n","seqfile_name: Gallus_gallus.GRCg7b.2bit\n","seqnames: c(%s)\n","circ_seqs: 'MT'\n","genome: GRCg7b\n","BSgenomeObjname: Ggallus\n","organism_biocview: Gallus_gallus\",\n","paste(dQuote(seqnames_data), collapse = \", \"))\n","\n","writeLines(seed_content, \"/content/drive/MyDrive/Bioinformatics/Colab_Lib/Saved_Files/GRCg7b.110/seed.txt\")"],"metadata":{"collapsed":true,"id":"B86-JT3NsiA-"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Check that output populated correctly"],"metadata":{"id":"PuHHOn-0wkGr"}},{"cell_type":"code","source":["!cat /content/drive/MyDrive/Bioinformatics/Colab_Lib/Saved_Files/GRCg7b.110/seed.txt"],"metadata":{"collapsed":true,"id":"arZ1nhdkttRd"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Forge package"],"metadata":{"id":"TtJDYU7yt90y"}},{"cell_type":"code","source":["!rm -rf /content/drive/MyDrive/Bioinformatics/Colab_Lib/Saved_Files/GRCg7b.110/BSgenome.Ggallus.ensembl.GRCg7b"],"metadata":{"id":"HC52wuALw5Be"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%R forgeBSgenomeDataPkg(\"/content/drive/MyDrive/Bioinformatics/Colab_Lib/Saved_Files/GRCg7b.110/seed.txt\", destdir = \"/content/drive/MyDrive/Bioinformatics/Colab_Lib/Saved_Files/GRCg7b.110\")"],"metadata":{"collapsed":true,"id":"N6s6BjKNt4f8"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Install newly forged package"],"metadata":{"id":"i8SPQQdLwL6g"}},{"cell_type":"code","source":["%%bash\n","cd /content/drive/MyDrive/Bioinformatics/Colab_Lib/Saved_Files/GRCg7b.110\n","R CMD build BSgenome.Ggallus.ensembl.GRCg7b"],"metadata":{"collapsed":true,"id":"56dqDzX06m-P"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","path <- \"/content/drive/MyDrive/Bioinformatics/Colab_Lib/Saved_Files/GRCg7b.110/BSgenome.Ggallus.ensembl.GRCg7b_1.0.0.tar.gz\"\n","install.packages(path, repos = NULL, type = \"source\", lib = \"/content/drive/MyDrive/Bioinformatics/Colab_Lib/R\")"],"metadata":{"id":"YqKICIL38yAY"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Generating Input Files"],"metadata":{"id":"bXPYWeYJZI6r"}},{"cell_type":"markdown","source":["Next, we need to generate four files that will be used as input for the main notebook:\n","\n","\n","1.   Seqinfo.csv: This stores important info about genomic sequences (names, lengths, circularity, etc)\n","2.   Gene Transfer Format (.gtf): Describes gene/transcript features of the reference genome.\n","3.   W.txt: a list of W-chromosome-linked genes.\n","4.   MT.txt: a list of mitochondria-linked genes.\n","\n"],"metadata":{"id":"cGk6LRCWZ9o-"}},{"cell_type":"markdown","source":["FIRST: SEQINFO. Get sequence info from earlier 2bit file:"],"metadata":{"id":"n5ru1voRa2XN"}},{"cell_type":"code","source":["%%R\n","twobit_file <- \"/content/drive/MyDrive/Bioinformatics/Colab_Lib/Saved_Files/GRCg7b.110/Gallus_gallus.GRCg7b.2bit\"\n","two_bit <- TwoBitFile(twobit_file)\n","seqinfo_data <- seqinfo(two_bit)\n","seqinfo_data <- Seqinfo(seqnames = seqnames(seqinfo_data),\n","  seqlengths = seqlengths(seqinfo_data), genome = \"GRCg7b\")"],"metadata":{"id":"Fy1pxjHrak9Z"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Create data frame and output to file:"],"metadata":{"id":"K1Cnb1-ubARU"}},{"cell_type":"code","source":["%%R\n","seqinfo_df <- data.frame(seqnames = seqnames(seqinfo_data),\n","  length = seqlengths(seqinfo_data),isCircular = seqnames(seqinfo_data) %in% c(\"MT\"))\n","csv_file <- \"/content/drive/MyDrive/Bioinformatics/Colab_Lib/Saved_Files/GRCg7b.110/GRCg7b.110.SeqInfo.csv\"\n","write.csv(seqinfo_df, file = csv_file,row.names = FALSE,quote = FALSE)"],"metadata":{"id":"s_rne5zHbHrC"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["NEXT: GTF (this is actually just a download and unzip)"],"metadata":{"id":"-su-EAqQdzLi"}},{"cell_type":"code","source":["%%R\n","gtf_url <- \"https://ftp.ensembl.org/pub/release-110/gtf/gallus_gallus/Gallus_gallus.bGalGal1.mat.broiler.GRCg7b.110.gtf.gz\"\n","gtf_zip_file <- \"/content/drive/MyDrive/Bioinformatics/Colab_Lib/Saved_Files/GRCg7b.110/GRCg7b.110.gtf.gz\"\n","gtf_file <- \"/content/drive/MyDrive/Bioinformatics/Colab_Lib/Saved_Files/GRCg7b.110/GRCg7b.110.gtf\"\n","download.file(gtf_url, destfile = gtf_zip_file)\n","R.utils::gunzip(gtf_zip_file, destname = gtf_file, remove=FALSE)"],"metadata":{"collapsed":true,"id":"KK65w-jBd8Xj"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["NEXT: W and MT files"],"metadata":{"id":"JN-Qvdxre9Pc"}},{"cell_type":"markdown","source":["Import GTF file, extract gene features"],"metadata":{"id":"S1-sTr0tfR9x"}},{"cell_type":"code","source":["%%R\n","gtf <- rtracklayer::import(gtf_file, format = \"gtf\")\n","genes <- gtf[gtf$type == \"gene\"]"],"metadata":{"id":"RAeZz9qee_pn"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Set preferred identifier (name or ID), remove any NAs:"],"metadata":{"id":"eKCmxaDMfbUu"}},{"cell_type":"code","source":["%%R\n","get_preferred_id <- function(granges_obj) {\n","  ids <- ifelse(!is.na(granges_obj$gene_name),\n","               granges_obj$gene_name,\n","               granges_obj$gene_id)\n","  ids[!is.na(ids)] }"],"metadata":{"id":"Tmo62z5vffYp"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Extract genes:"],"metadata":{"id":"ZhpeigpWgDqR"}},{"cell_type":"code","source":["%%R\n","# For W chromosome genes:\n","genes_W <- genes[seqnames(genes) == \"W\"]\n","gene_ids_W <- get_preferred_id(genes_W)\n","# For MT chromosome genes:\n","genes_MT <- genes[seqnames(genes) == \"MT\"]\n","gene_ids_MT <- get_preferred_id(genes_MT)"],"metadata":{"id":"A-VvIn4xfo4N"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Write output files:"],"metadata":{"id":"nZo7iD67gBVE"}},{"cell_type":"code","source":["%%R\n","output_W <- \"/content/drive/MyDrive/Bioinformatics/Colab_Lib/Saved_Files/GRCg7b.110/GRCg7b.110.W.txt\"\n","output_MT <- \"/content/drive/MyDrive/Bioinformatics/Colab_Lib/Saved_Files/GRCg7b.110/GRCg7b.110.MT.txt\"\n","write.table(gene_ids_W, file = output_W,\n","            row.names = FALSE, col.names = FALSE, quote = FALSE)\n","write.table(gene_ids_MT, file = output_MT,\n","            row.names = FALSE, col.names = FALSE, quote = FALSE)"],"metadata":{"id":"7fCyDxNQgGhn"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["List contents of working directory:"],"metadata":{"id":"vjrvquWZgrMy"}},{"cell_type":"code","source":["!ls /content/drive/MyDrive/Bioinformatics/Colab_Lib/Saved_Files/GRCg7b.110"],"metadata":{"id":"zr5Ht8WGgucG"},"execution_count":null,"outputs":[]}],"metadata":{"colab":{"provenance":[],"machine_shape":"hm","authorship_tag":"ABX9TyMmgQqLicyhxNOPTbweSVE4"},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"name":"python"}},"nbformat":4,"nbformat_minor":0}